libsnark/gadgetlib1/gadgets/curves/scalar_multiplication.hpp

```diff
@@ -10,12 +10,295 @@
 #define LIBSNARK_GADGETLIB1_GADGETS_CURVE_SCALAR_MULTIPLICATION_HPP_
 
 #include "libsnark/gadgetlib1/gadget.hpp"
+#include "libsnark/gadgetlib1/gadgets/basic_gadgets.hpp"
+#include "libsnark/gadgetlib1/gadgets/pairing/pairing_params.hpp"
 
+#include <libff/algebra/curves/public_params.hpp>
 #include <memory>
 
 namespace libsnark
 {
 
+/// Wrapper for group variable, adding a flag that indicates whether the
+/// variable is the identity.
+///
+/// For use in 2 possible configurations:
+///
+///   1) Construct with variable_or_identity(pb, annotation_prefix), which
+///      internally allocates variables for the group element and flag.
+///      Values can later be set via generate_r1cs_witness() calls.
+///
+///   2) Construct using a group element value, which defines values as linear
+///      combinations (no variable allocation). generate_r1cs_witness() calls
+///      should NOT be made to such objects.
+template<typename ppT, typename groupT, typename groupVariableT>
+class variable_or_identity : public gadget<libff::Fr<ppT>>
+{
+public:
+    using FieldT = libff::Fr<ppT>;
+
+    groupVariableT value;
+    pb_linear_combination<FieldT> is_identity;
+
+    // Internally allocates the variable and flag, to be set later via the
+    // generate_r1cs_witness calls.
+    variable_or_identity(
+        protoboard<FieldT> &pb, const std::string &annotation_prefix);
+
+    // Initialize the variable and flag with constants. Do not try to call
+    // generate_r1cs_witness().
```

Comment:
What happens if `generate_r1cs_witness()` is called after this constructor is used?

Reply:
Good question.  I guess it really depends on exactly how this variable is used in the rest of the circuit.  If these values are set to a constant with this constructor, and the rest of the circuit constructed based on that, and then the value overwritten by calling `generate_r1cs_witness()`, I would expect some constraints in the system to no longer be satisfied, or possibly the circuit could be satisfied, but inputs to other gadgets may no longer be as expected.

I guess if one provided the same value as witness data it would not fail.  But as a principle, if a constant is assigned when the circuit is constructed, it defines coefficients of `var[0]` (the implicit variable with value `1`) to be used in other parts of the circuit.  It is inconcistent to then try to assign witness data to this value.

The same holds for the underlying `G1_variable` (and possibly other gadgets, such as extension field elements).  I have been wondering whether it could make sense to use different types for the constant cases, since we should be able to know at compile time which will be used, and then statically prevent witness data being assigned.  However, it would take some thought to make it fit nicely with existing gadgets.